features/cellular/framework/targets/UBLOX/N2XX/UBLOX_N2XX_CellularContext.cpp

```diff
@@ -0,0 +1,50 @@
+/*
+ * Copyright (c) 2018, Arm Limited and affiliates.
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "UBLOX_N2XX_CellularContext.h"
+#include "UBLOX_N2XX_CellularStack.h"
+#include "APN_db.h"
+
+namespace mbed {
+
+UBLOX_N2XX_CellularContext::UBLOX_N2XX_CellularContext(ATHandler &at, CellularDevice *device, const char *apn) :
+    AT_CellularContext(at, device, apn)
+{
+}
+
+UBLOX_N2XX_CellularContext::~UBLOX_N2XX_CellularContext()
+{
+}
+
+NetworkStack *UBLOX_N2XX_CellularContext::get_stack()
+{
+    if (!_stack) {
+        _stack = new UBLOX_N2XX_CellularStack(_at, _cid, _ip_stack_type);
```

Comment:
What is our view on being "Thread Safe"? 
Is this "by design" - in that the other libraries safely call this get_stack() function, or do we just state these drivers are not Thread Safe?

Reply:
Not sure but seems like these are thread safe, as Mutex is being locked before using AT interface..

Reply:
It's not about the AT interface though. If two threads call get_stack(), you might end up with two UBLOX_N2XX_CellularStack objects. 

I know this code is trying to provide the same instance back if already created, but this code isn't thread safe. 

Have a look at double check locking when creating the singleton object - but there are issues in how to implement this as compilers may want to optimise it: https://www.aristeia.com/Papers/DDJ_Jul_Aug_2004_revised.pdf

Reply:
@RobMeades  - What should we do here? be squeaky clean and try to fix it? or just say "It's not thread safe"

Reply:
According to my understanding, NetworkInterface class is exposed to user application so user dont have direct access to these cellular classes. get_stack() like function are for internal mbed usage and mbed is making sure for singleton instance.
https://github.com/u-blox/mbed-os/blob/df9ac85cbb38d8f06380b339398e73f968e3ba0a/features/cellular/framework/API/CellularDevice.h#L53

Reply:
OK - so this is "by design" of using static and hoping the compiler does the right thing.